cdc/redo/manager.go

```diff
@@ -460,256 +357,152 @@ func (m *ManagerImpl) flushLog(
-						rtsMap.ReplaceOrInsert(cache.span, cache.resolvedTs)
-					}
-				default:
-					log.Panic("redo manager receives unknown event type")
-				}
-				workTimeSlice += time.Since(startToHandleEvent)
-			case now := <-overseerTicker.C:
-				busyRatio := int(workTimeSlice.Seconds() / now.Sub(startToWork).Seconds() * 1000)
-				m.metricRedoWorkerBusyRatio.Add(float64(busyRatio))
-				startToWork = now
-				workTimeSlice = 0
-			case err = <-logErrCh:
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-ticker.C:
+			m.flushLog(ctx, handleErr, &workTimeSlice)
+		case event, ok := <-m.logBuffer.Out():
+			if !ok {
+				return nil // channel closed
 			}
+			err = m.handleEvent(ctx, event, &workTimeSlice)
+		case now := <-overseerTicker.C:
+			busyRatio := int(workTimeSlice.Seconds() / now.Sub(startToWork).Seconds() * 1000)
+			m.metricRedoWorkerBusyRatio.Add(float64(busyRatio))
+			startToWork = now
+			workTimeSlice = 0
+		case err = <-logErrCh:
 		}
 		if err != nil {
 			log.Warn("redo manager writer meets write or flush fail",
-				zap.String("namespace", m.changeFeedID.Namespace),
-				zap.String("changefeed", m.changeFeedID.ID),
+				zap.String("namespace", m.cfg.ChangeFeedID.Namespace),
+				zap.String("changefeed", m.cfg.ChangeFeedID.ID),
 				zap.Error(err))
-			break
+			return err
 		}
 	}
-
-	// NOTE: the goroutine should never exit until the error is put into errCh successfully
-	// or the context is canceled.
-	select {
-	case <-ctx.Done():
-	case errCh <- err:
-	}
 }
 
-func (m *ManagerImpl) bgGC(ctx context.Context) {
-	ticker := time.NewTicker(time.Duration(defaultGCIntervalInMs) * time.Millisecond)
-	defer ticker.Stop()
-	for {
-		select {
-		case <-ctx.Done():
-			log.Info("redo manager GC exits as context cancelled",
-				zap.String("namespace", m.changeFeedID.Namespace),
-				zap.String("changefeed", m.changeFeedID.ID))
-			return
-		case <-ticker.C:
-			ckpt := atomic.LoadUint64(&m.checkpointTs)
-			if ckpt == 0 {
-				continue
-			}
-			log.Debug("redo manager GC is triggered",
-				zap.Uint64("checkpointTs", ckpt),
-				zap.String("namespace", m.changeFeedID.Namespace),
-				zap.String("changefeed", m.changeFeedID.ID))
-			err := m.writer.GC(ctx, ckpt)
-			if err != nil {
-				log.Warn("redo manager log GC fail",
-					zap.String("namespace", m.changeFeedID.Namespace),
-					zap.String("changefeed", m.changeFeedID.ID), zap.Error(err))
-			}
-		}
-	}
-}
-
-func (m *ManagerImpl) withLock(action func(m *ManagerImpl) error) error {
+func (m *logManager) withLock(action func(m *logManager) error) error {
 	m.rwlock.RLock()
 	defer m.rwlock.RUnlock()
 	if atomic.LoadInt32(&m.closed) != 0 {
-		return cerror.ErrRedoWriterStopped.GenWithStack("redo manager is stopped")
+		return errors.ErrRedoWriterStopped.GenWithStack("redo manager is closed")
 	}
 	return action(m)
 }
+
+func (m *logManager) close() {
+	m.rwlock.Lock()
+	defer m.rwlock.Unlock()
+	atomic.StoreInt32(&m.closed, 1)
```

Comment:
Where to close the background goroutine?

Reply:
The timing here is to exit the background goroutine first and close manager:

![image](https://user-images.githubusercontent.com/61726649/220869887-62b1c5ae-30c3-45d8-9bbc-b1a32fdd59e1.png)